File: src/main/java/dev/slne/discord/config/BotConfig.java
package dev.slne.discord.config;

import dev.slne.discord.config.discord.DiscordBotConfig;
import dev.slne.discord.config.discord.GuildConfig;
import dev.slne.discord.config.role.RoleConfig;
import dev.slne.discord.config.ticket.TicketConfig;
import lombok.Getter;
import lombok.ToString;
import org.jetbrains.annotations.ApiStatus;
import org.spongepowered.configurate.CommentedConfigurationNode;
import org.spongepowered.configurate.objectmapping.ConfigSerializable;
import org.spongepowered.configurate.yaml.YamlConfigurationLoader;

import java.nio.file.Path;
import java.util.Map;

/**
 * The interface Bot config.
 */
@ApiStatus.Internal
@ConfigSerializable
@ToString
@Getter
public class BotConfig {
	private static BotConfig instance;

	private Map<String, GuildConfig> guildConfig;
	private Map<String, RoleConfig> roleConfig;

	private DiscordBotConfig discordBotConfig;
	private TicketConfig ticketConfig;

	/**
	 * Instantiates a new Bot config.
	 */
	public BotConfig() {
	}

	/**
	 * Gets config.
	 *
	 * @return the config
	 */
	public static BotConfig getConfig() {
		if (instance != null) {
			return instance;
		}

		loadConfig();

		return instance;
	}

	/**
	 * Load config.
	 */
	public static void loadConfig() {
		YamlConfigurationLoader loader = YamlConfigurationLoader.builder().path(Path.of("data/config.yml")).build();

		try {
			CommentedConfigurationNode node = loader.load();

			if (node == null) {
				throw new IllegalStateException("Could not load configuration");
			}

			instance = node.get(BotConfig.class);
		} catch (Exception e) {
			System.err.println("An error occurred while loading this configuration: " + e.getMessage());

			if (e.getCause() != null) {
				e.getCause().printStackTrace();
			}
		}
	}

}

File: src/main/kotlin/dev/slne/discord/config/BotConfig.kt
package dev.slne.discord.config

import com.charleskorn.kaml.Yaml
import com.charleskorn.kaml.decodeFromStream
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.jetbrains.annotations.ApiStatus
import kotlin.io.path.Path
import kotlin.io.path.inputStream
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

@ApiStatus.Internal
@Serializable
data class BotConfig(
    @SerialName("bot-token") val botToken: String,
    val database: DatabaseConfig,
    val cooldown: CooldownConfig = CooldownConfig()
) {
    @Serializable
    data class CooldownConfig(
        @SerialName("faq-cooldown")
        val faqCooldown: Duration = 10.seconds
    )
}

val botConfig by lazy {
    Path("data/config.yml").inputStream().use { Yaml.default.decodeFromStream<BotConfig>(it) }
}